CI on windows is broken: PermissionError

See: https://github.com/huggingface/datasets/actions/runs/7104781624/job/19340572394
```
FAILED tests/test_load.py::test_loading_from_the_datasets_hub - NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpfcnps56i\\hf-internal-testing___dataset_with_script\\default\\0.0.0\\c240e2be3370bdbd\\dataset_with_script-train.arrow'
```

--- a/src/datasets/dataset_dict.py
+++ b/src/datasets/dataset_dict.py
@@ -56,6 +56,17 @@ class DatasetDict(dict):
                     f"All datasets in `DatasetDict` should have the same features but features for '{item_a[0]}' and '{item_b[0]}' don't match: {item_a[1].features} != {item_b[1].features}"
                 )
 
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # Here `del` is used to del the pyarrow tables. This properly closes the files used for memory mapped tables
+        for dataset in self.values():
+            if hasattr(dataset, "_data"):
+                del dataset._data
+            if hasattr(dataset, "_indices"):
+                del dataset._indices
+
     def __getitem__(self, k) -> Dataset:
         if isinstance(k, (str, NamedSplit)) or len(self) == 0:
             return super().__getitem__(k)

--- a/tests/test_load.py
+++ b/tests/test_load.py
@@ -1313,10 +1313,9 @@ def test_load_dataset_with_unsupported_extensions(text_dir_with_unsupported_exte
 @pytest.mark.integration
 def test_loading_from_the_datasets_hub():
     with tempfile.TemporaryDirectory() as tmp_dir:
-        dataset = load_dataset(SAMPLE_DATASET_IDENTIFIER, cache_dir=tmp_dir)
-        assert len(dataset["train"]) == 2
-        assert len(dataset["validation"]) == 3
-        del dataset
+        with load_dataset(SAMPLE_DATASET_IDENTIFIER, cache_dir=tmp_dir) as dataset:
+            assert len(dataset["train"]) == 2
+            assert len(dataset["validation"]) == 3
 
 
 @pytest.mark.integration

